README.md

```diff
@@ -2592,5 +2592,298 @@ TODO
+- *Present the result of the calculation to the verifier through a hint*, then show said result indeed satisfies the relevant condition that makes it the actual result.
+
+To make this less abstract, let's show two examples.
+
+##### Square root
+
+Let's say the calculation in question is to compute the square root of a number `x`. The two ways to do it then become:
+
+- Write the usual square root algorithm in Cairo to compute `sqrt(x)`.
+- Write a hint that computes `sqrt(x)`, then immediately after calling the hint show __in Cairo__ that `(sqrt(x))^2 = x`.
+
+The second approach is exactly what the `sqrt` function in the Cairo common library does:
+
+```
+// Returns the floor value of the square root of the given value.
+// Assumptions: 0 <= value < 2**250.
+func sqrt{range_check_ptr}(value) -> felt {
+    alloc_locals;
+    local root: felt;
+
+    %{
+        from starkware.python.math_utils import isqrt
+        value = ids.value % PRIME
+        assert value < 2 ** 250, f"value={value} is outside of the range [0, 2**250)."
+        assert 2 ** 250 < PRIME
+        ids.root = isqrt(value)
+    %}
+
+    assert_nn_le(root, 2 ** 125 - 1);
+    tempvar root_plus_one = root + 1;
+    assert_in_range(value, root * root, root_plus_one * root_plus_one);
+
+    return root;
+}
+```
+
+If you read it carefully, you'll see that the hint in this function computes the square root in python, then this line
+
+```
+assert_in_range(value, root * root, root_plus_one * root_plus_one);
+```
+
+asserts __in Cairo__ that `(sqrt(x))^2 = x`.
+
+This is done this way because it is much cheaper, in terms of the generated trace (and thus proving time), to square a number than compute its square root.
+
+Notice that the last assert is absolutely mandatory to make this safe. If you forget to write it, the square root calculation does not get proven, and anyone could convince the verifier that the result of `sqrt(x)` is any number they like.
+
+##### Linear search turned into an O(1) lookup
+
+This example is taken from the [Cairo documentation](https://docs.cairo-lang.org/0.12.0/hello_cairo/program_input.html).
+
+Given a list of `(key, value)` pairs, if we want to write a `get_value_by_key` function that returns the value associated to a given key, there are two ways to do it:
+
+- Write a linear search in Cairo, iterating over each key until you find the requested one.
+- Do that exact same linear search *inside a hint*, find the result, then show that the result's key is the one requested.
+
+Again, the second approach makes the resulting trace and proving much faster, because it's just a lookup; there's no linear search. Notice this only applies to proving, the VM has to execute the hint, so there's still a linear search when executing to generate the trace. In fact, the second approach is more expensive for the VM than the first one. It has to do both a linear search and a lookup. This is a tradeoff in favor of proving time.
```

Comment:
```suggestion
Again, the second approach makes the resulting trace and proving much faster, because it's just a lookup; there's no linear search. Notice this only applies to proving, the VM has to execute the hint, so there's still a linear search when executing to generate the trace. In fact, the second approach is more expensive for the VM than the first one. It has to do both a linear search and a lookup. *This is a tradeoff in favor of proving time.*
```

That was something important for me to understand why hints existed